tfhe/src/integer/client_key/mod.rs

```diff
@@ -256,26 +268,61 @@ impl ClientKey {
     /// assert_eq!(msg, dec);
     /// ```
     pub fn decrypt_radix_without_padding(&self, ctxt: &RadixCiphertext) -> u64 {
-        let mut result = 0_u64;
-        let mut shift = 1_u64;
-        let modulus = self.parameters().message_modulus.0 as u64;
-        for c_i in ctxt.blocks.iter() {
-            // decrypt the component i of the integer and multiply it by the radix product
-            let block_value = self
-                .key
-                .decrypt_message_and_carry_without_padding(c_i)
-                .wrapping_mul(shift);
-
-            // update the result
-            result = result.wrapping_add(block_value);
+        let mut res = 0u64;
+        self.decrypt_radix_into_words(
+            ctxt,
+            res.as_words_mut(),
+            crate::shortint::ClientKey::decrypt_message_and_carry_without_padding,
+        );
+        res
+    }
 
-            // update the shift for the next iteration
-            shift = shift.wrapping_mul(modulus);
+    /// Decrypts a ciphertext in radix decomposition into 64bits
+    ///
+    /// The words are assumed to be in little endian order.
+    pub fn decrypt_radix_into_words<F>(
+        &self,
+        ctxt: &RadixCiphertext,
+        clear_words: &mut [u64],
+        decrypt_block: F,
+    ) where
+        F: Fn(&crate::shortint::ClientKey, &crate::shortint::Ciphertext) -> u64,
+    {
+        #[cfg(target_endian = "little")]
+        let clear_words_iter = clear_words.iter_mut();
+        #[cfg(target_endian = "big")]
+        let clear_words_iter = clear_words.iter_mut().rev();
```

Comment:
well it has an impact on native types and our types I believe, depending on how we store things we may even have mixed endianness in the u256 🤔

Reply:
not sure i follow, If we use `[u64;4]` or `[u128;2]` we will end up with mixed endianness in one endian and not the other unless we do someting to have the "word" order match the endianness, but that can be made to work with both cases

Reply:
ah ok after re-reading it should work ok for u128 as you reverse the iterator

Reply:
yes you I right that in big-endian U256::as_words won't result in words in the right order (even after the reverse)

We can fix this by:
1) making the low/ high order of the u128 used within the U256 depend on the target endianness
2) make the as_words trait return a `Vec<u64>` to let the implementation reorder words
3) use 4 u64 and still use the reverse

Reply:
I would maybe look at the 4 u64 if it makes sense given the choice made for the ntt

I'm not sure solution 1 with different u128 order is good idea if there is serialization involved

Or we just say "we don't support big endian" I don't know how bad of a choice that would be

Solution 2 is a bit annoying but worst case it can work

Reply:
If we use 4 u64 we will still have to chose if the u64[0] is the MSB or LSB and then have to introduce a "reversing" at some point

Reply:
right, or choose an order dependent on the endianness and manage ordering for serialization as was done by Sarah for the FFT structs 🤔

Reply:
I did a refactor that should make things work properly on big endians (U256 changed + ClearWord trait changed)